cmake: use sdk-ng built toolchain for x86_64

dcpleung opened the pull request:
This adds the necessary bits to utilize the x86_64 toolchain
built by sdk-ng for x86_64 when toolchain variant is either
zephyr or xtools. This allows decoupling the builds from
the host toolchain.

This needs https://github.com/zephyrproject-rtos/sdk-ng/pull/56 to be merged first.

aescolar reviewed (dismissed):
About the PR and commit title, they seem misleading: (correct me if I'm wrong, but) here you are "Enabling", not changing anything for those who keep ZEPHYR_TOOLCHAIN_VARIANT=zephyr , or?
(meaning nobody, including CI, would necessarily realize about this PR being merged (ignoring the changes in the x64 linker options))

Another nit about the description: this would be the case not only native_posix but all POSIX arch boards. ( nrf52_bsim exists ;) , even that many people only remembers about native_posix or confuse them )

dcpleung commented:
Realized after waking up that this would probably not work on Windows so remove the native_posix part, and just targeting x86_64.

@aescolar This is just for xtools and does not affect the usage of official SDK, in other words, only used when `ZEPHYR_TOOLCHAIN_VARIANT=xtools`.

andyross commented on cmake/generic_toolchain.cmake:
```diff
@@ -47,7 +47,10 @@ set(ZEPHYR_TOOLCHAIN_VARIANT ${ZEPHYR_TOOLCHAIN_VARIANT} CACHE STRING "Zephyr to
+# or x86_64 (but not using xtools)
+if((${ARCH} STREQUAL "posix") OR
+   ((${ARCH} STREQUAL "x86_64")
+    AND (NOT ${ZEPHYR_TOOLCHAIN_VARIANT} STREQUAL "xtools")))
```
Once the x86_64 compiler is in the SDK, we can actually remove this bit and just use ZEPHYR_TOOLCHAIN_VARIANT=host as it was originally intended.

Maybe revert the sense here and force the toolchain to host only if it started at "zephyr" to begin with, for greater compatibility?  Probably not a big deal.

dcpleung replied:
Once the SDK has the x86_64 toolchain, we can remove the the "x86_64 and !xtools" part as x86_64 will be built using the toolchain in SDK.

marc-hb replied:
> as x86_64 will be built using the toolchain in SDK.

again this seems to imply that the number of toolchain options will be reduced. Is this the case?

dcpleung replied:
The intention is to test the sdk-ng built x86_64 toolchain before we actually include it in the SDK, hence enabling it for xtools, for now. This also allows developers to have a common base for comparison if there is a bug report. Once we have the x86_64 toolchain in the official SDK, the host toolchain will only be used for when ARCH==posix.

Does this answer your question?

marc-hb replied:
> Once we have the x86_64 toolchain in the official SDK, the host toolchain will only be used for when ARCH==posix.

AFAIK the choice of the toolchain is made by the end-user with the ZEPHYR_TOOLCHAIN_VARIANT variable. Do you plan to remove that choice?

dcpleung replied:
No. The change inside cmake/toolchain/xtools will stay.

Or, do you want to use the host toolchain to build for x86_64 even if `ZEPHYR_TOOLCHAIN_VARIANT=xtools`?

andyross replied:
@marc-hb: it's sort of the converse.  Right now, there is no SDK support for x86_64 and posix, so we effectively force ZEPHYR_TOOLCHAIN_VARIANT to be "host" for those platforms even if it started out as "zephyr".

marc-hb replied:
> we effectively force ZEPHYR_TOOLCHAIN_VARIANT to be "host" for those platforms even if it started out as "zephyr".

Thanks, I didn't expect such a hack. I guess it was added to keep CI simple while adding some coverage?

So because the current (and too short) commit message/description assumes knowledge of this hack it almost sounds like it's adding such a restriction/hack when it will actually allow removing one...

andyross reviewed (approved):
Nitpick on the sense of the toolchain variant string, looks clean to me.

marc-hb commented:
>  This allows decoupling the builds from the host toolchain.

Agree with @aescolar that this patch has a confusing top-level description. I think you mean: "This allows building x86_64 with new toolchain FOO in addition to the existing host and zephyr toolchains"
There's no feature change for the existing toolchains, correct?

marc-hb commented:
BTW https://github.blog/2019-02-14-introducing-draft-pull-requests/

dcpleung commented:
Arr... I copied the summary from the patch in sdk-ng. Updated the title of this PR to the actual summary line in the commit.

nashif commented on boards/x86_64/qemu_x86_64/qemu_x86_64.yaml:
```diff
@@ -6,5 +6,6 @@ arch: x86_64
 toolchain:
   - zephyr
   - host
+  - xtools
```
why xtools? This should work with the zephyr variant, no?

dcpleung replied:
Updated the patch to use the newly built x86_64 toolchain for when toolchain variant is zephyr. This is for testing only with artifacts from CI. This needs to be updated for new SDK release.

dcpleung replied:
Should work for both zephyr and xtools now.

andrewboie commented on cmake/toolchain/zephyr/0.10.0/target.cmake:
```diff
@@ -28,5 +29,12 @@ if("${ARCH}" STREQUAL "xtensa")
+elseif("${ARCH}" STREQUAL "x86_64")
+  list(APPEND TOOLCHAIN_C_FLAGS -mx32)
+  list(APPEND TOOLCHAIN_LD_FLAGS -mx32)
+  set(TOOLCHAIN_HAS_NEWLIB OFF CACHE BOOL "True if toolchain supports newlib")
```
can we put a build of newlib in the SDK for x86_64? AFAIK the network stack currently requires newlib, or at least the tests do

dcpleung replied:
newlib doesn't build correctly with `-mx32` on my machine, which results in linking error for `zephyr.elf`. Also, under crosstool-ng, newlib does not have multilib support. This means we could only build for `x86`, `x32` or `x86_64` but not all three. Maybe if we only do `x86` and `x86_64` (without `x32`), newlib might build correctly. We can try again for newlib if we drop `x32`.

andrewboie replied:
oof. the kernel is very non 64-bit clean, I think we're going to be using x32 for a while. I agree with @andyross it may be worth investigating what's breaking x32 newlib build, agree its probably something trivial.

andyross commented:
Or we could try to fix whatever the bug is with x32 and newlib.  It's probably something simple (like predicating sizes on __x86_64 instead of checking the __*_WIDTH__ macros, etc...)

dcpleung commented:
> Or we could try to fix whatever the bug is with x32 and newlib. It's probably something simple (like predicating sizes on `__x86_64` instead of checking the `__*_WIDTH__` macros, etc...). Shouldn't hold up this patch, of course.

I looked into it for about a week or so. The issue I had was due to combination of crosstool-ng and newlib's build system. Somehow it resulted in libraries for exactly one architecture (even for gcc itself). Depending on how the "architectures to support list" is ordered, it would either result in all straight 64-bit libraries, or straight 32-bit libraries. So `zephyr.elf` would never be able to link correctly (no x32), and bootstub might compile/link if there are 32-bit libraries laying around.

dcpleung committed:
cmake: use sdk-ng built toolchain for x86_64

This adds the necessary bits to utilize the x86_64 toolchain
built by sdk-ng for x86_64 when toolchain variant is either
zephyr or xtools. This allows decoupling the builds from
the host toolchain.

Newlib is also available with this toolchain so remove
the Kconfig restriction on CONFIG_NEWLIB_LIBC.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
x86_64: minimally preparing for enabling newlib

The libc hooks for Newlib requires CONFIG_SRAM_SIZE and
the symbol "_end" at the end of memory. This is in preparation
for enabling Newlib for x86_64.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Got newlib to work. This needs the latest patch @ https://github.com/zephyrproject-rtos/sdk-ng/pull/56

dcpleung commented:
https://github.com/zephyrproject-rtos/sdk-ng/pull/56 is merged. So this is no longer blocked.

galak commented:
> [zephyrproject-rtos/sdk-ng#56](https://github.com/zephyrproject-rtos/sdk-ng/pull/56) is merged. So this is no longer blocked.

it should be blocked til we have a release and use that release in CI.

dcpleung committed:
cmake: bump min SDK version 0.10.1

This new SDK:
() Fixes an issue with i586 toolchain where no coverage data
   would be produced;
() Adds a new x86_64 toolchain for building x86_64
   targets, decoupling x86_64 builds from host toolchain;
() Includes MIPS toolchain;
() Reverts bossa to older version to fix flashing issues;
() Turns on multilib support for RISC-V; and,
() Updates OpenOCD for TI and some ARC fixes.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
SDK 0.10.1 has been released, and this PR has been updated to use the new SDK.

SebastianBoe commented on arch/x86_64/CMakeLists.txt:
```diff
@@ -11,4 +11,7 @@ add_subdirectory(core)
-zephyr_link_libraries(-mx32)
+zephyr_link_libraries(
+    -mx32
+    -Wl,-melf32_x86_64
```
For portability, "-Wl" should be replaced with ${LINKERFLAGPREFIX} (after verifying that LINKERFLAGPREFIX is intialized before use).

dcpleung replied:
Replaced `-Wl` with `${LINKERFLAGPREFIX}`. The flag should have been defined as these run during the last few steps of the build process (or it would have error-ed out earlier). I don't see anywhere else actually checking for it being defined though.

marc-hb commented:
Now that this is merged I started using qemu_x86_64. There is a fairly frequent assert in `handler.log` and only for qemu_x86_64:
```
ASSERTION FAIL [log_backend_count_get() > 0] @ ZEPHYR_BASE/subsys/logging/log_core.c:828
```

A bit worryingly, all the tests PASS anyway. Now to be fair sanitycheck would have a hard time catching many of these ASSERT because the way they're logged is racy. Screenshot of one of the worst I've seen:

qemu_x86_64/tests/posix/fs/portability.posix/handler.log

```
SeaBIOS (version rel-1.12.0-0-ga698c8995f-prebuilt.qemu.org)
Booting from ROM..Running test suite posix_fs_test
===================ASSE==================RTION =======FAIL===== [log_==backend_count_get(===) > 0] @ Z==EPHYR_BASE/subsy===s/logging/log_core=====.c:===
starting test - test_fs_mount
828
```

cc @npitre just in case it's a 64bit issue.

npitre commented:
I'll try to reproduce here and have a look.
This might go to next week though.

npitre commented:
On Thu, 27 Jun 2019, Marc wrote:

> cc @npitre just in case it's a 64bit issue.

Well... log_backend_count_get() is defined as:

    static inline int log_backend_count_get(void)
    {
            return __log_backends_end - __log_backends_start;
    }

Those are defined in include/linker/common-rom.ld:

        SECTION_DATA_PROLOGUE(log_backends_sections,,)
        {
                __log_backends_start = .;
                KEEP(*(".log_backends"));
                __log_backends_end = .;
        } GROUP_LINK_IN(ROMABLE_REGION)

So I hardly see how 64bit-ness could have had any influence here.

Maybe you could have a look in your zephyr.map file to see where those 
symbols are actually located and if the section is really empty?

dcpleung commented:
That assertion is correct since there really is no backend defined. `CONFIG_LOG_BACKEND_UART=y` is defined for `qemu_x86`, but nothing for `qemu_x86_64`.

marc-hb commented:
> So I hardly see how 64bit-ness could have had any influence here.

Indeed not, thanks for your time looking at this.  Issue #17555 filed by @andrewboie
